backend/metadata_writer/src/metadata_writer.py

```diff
@@ -0,0 +1,756 @@
+metadata_service_port = int(os.environ.get('METADATA_SERVICE_SERVICE_PORT', 8080))
+
+
+mlmd_connection_config = metadata_store_pb2.MetadataStoreClientConfig(
+    host=metadata_service_host,
+    port=metadata_service_port,
+)
+mlmd_store = metadata_store.MetadataStore(mlmd_connection_config)
+print("Connected to the metadata store")
+
+#%%
+
+ARGO_OUTPUTS_ANNOTATION_KEY = 'workflows.argoproj.io/outputs'
+ARGO_TEMPLATE_ANNOTATION_KEY = 'workflows.argoproj.io/template'
+KFP_COMPONENT_SPEC_ANNOTATION_KEY = 'pipelines.kubeflow.org/component_spec'
+METADATA_EXECUTION_ID_LABEL_KEY = 'pipelines.kubeflow.org/metadata_execution_id'
+METADATA_CONTEXT_ID_LABEL_KEY = 'pipelines.kubeflow.org/metadata_context_id'
+METADATA_ARTIFACT_IDS_ANNOTATION_KEY = 'pipelines.kubeflow.org/metadata_artifact_ids'
+METADATA_INPUT_ARTIFACT_IDS_ANNOTATION_KEY = 'pipelines.kubeflow.org/metadata_input_artifact_ids'
+METADATA_OUTPUT_ARTIFACT_IDS_ANNOTATION_KEY = 'pipelines.kubeflow.org/metadata_output_artifact_ids'
+
+ARGO_WORKFLOW_LABEL_KEY = 'workflows.argoproj.io/workflow'
+ARGO_COMPLETED_LABEL_KEY = 'workflows.argoproj.io/completed'
+METADATA_WRITTEN_LABEL_KEY = 'pipelines.kubeflow.org/metadata_written'
+
+
+#%%
+
+def cleanup_pods():
+    for pod in k8s_api.list_namespaced_pod(namespace=namespace_to_watch, label_selector=ARGO_WORKFLOW_LABEL_KEY).items:
+        patch_pod_metadata(
+            namespace=pod.metadata.namespace,
+            pod_name=pod.metadata.name,
+            patch={
+                'annotations': {
+                    METADATA_ARTIFACT_IDS_ANNOTATION_KEY: None,
+                    METADATA_INPUT_ARTIFACT_IDS_ANNOTATION_KEY: None,
+                    METADATA_OUTPUT_ARTIFACT_IDS_ANNOTATION_KEY: None,
+                },
+                'labels': {
+                    METADATA_EXECUTION_ID_LABEL_KEY: None,
+                    METADATA_WRITTEN_LABEL_KEY: None,
+                    METADATA_CONTEXT_ID_LABEL_KEY: None,
+                }
+            },
+        )
+#%%
+
+#cleanup_pods()
+
+
+#%%
+
+def output_name_to_argo(name: str) -> str:
+    import re
+    return re.sub('-+', '-', re.sub('[^-0-9a-z]+', '-', name.lower())).strip('-')
+
+
+def argo_artifact_to_uri(artifact: dict) -> str:
+    if 's3' in artifact:
+        s3_artifact = artifact['s3']
+        return 'https:/artifacts/get?source=minio&bucket={bucket}&key={key}'.format(
+            bucket=s3_artifact.get('bucket', ''),
+            key=s3_artifact.get('key', ''),
+        )
+    elif 'raw' in artifact:
+        return None
+    else:
+        return None
+
+
+def is_tfx_pod(pod) -> bool:
+    main_containers = [container for container in pod.spec.containers if container.name == 'main']
+    if len(main_containers) != 1:
+        return False
+    main_container = main_containers[0]
+    return main_container.command and main_container.command[-1].endswith('tfx/orchestration/kubeflow/container_entrypoint.py')
```

Comment:
Just FYI, in TFX docker file the default entrypoint is run_executor.py
`https://github.com/tensorflow/tfx/blob/fa53f0da94c29963aa4fd488b54c858398e391fb/tfx/tools/docker/Dockerfile#L81`

No need to change for now, but we might want to make this configurable in the future.

Reply:
Thanks for the observation! I'll parametrize it.

Reply:
AFAIK, For now, all pods created by TFX using the KubeflowLauncher (both python and container) are using the "container_entrypoint.py".